sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala

```diff
@@ -329,17 +337,26 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
    *
    * This is modeled after `RDD.take` but never runs any job locally on the driver.
    */
-  def executeTake(n: Int): Array[InternalRow] = {
+  def executeTake(n: Int): Array[InternalRow] = executeTakeSeqView(n)._2.force
+
+  /**
+   * Runs this query returning the tuple of the row count and the SeqView of first `n` rows.
+   *
+   * This is modeled to execute decodeUnsafeRows lazily to reduce peak memory usage of
+   * decoding rows. Only compressed byte arrays consume memory after return.
+   */
+  private[spark] def executeTakeSeqView(
```

Comment:
You'd be better to separate this PR into two parts you proposed in the PR description.

Reply:
Do you mean to separate this PR by sql part and thriftserver part?